clouddrift/wavelet.py

```diff
@@ -0,0 +1,801 @@
+    function. Such wavelet can be provided by the function ``morse_wavelet``.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Real- or complex-valued signals.
+    wavelet : np.ndarray
+        A suite of time-domain wavelets, typically returned by the function ``morse_wavelet``.
+        The length of the time axis of the wavelets must be the last one and matches the
+        length of the time axis of x. The other dimensions (axes) of the wavelets (orders and frequencies) are
+        typically organized as orders, frequencies, and time, unless specified by optional arguments freq_axis and order_axis.
+        The normalization of the wavelets is assumed to be "bandpass", if not use kwarg normalization="energy", see ``morse_wavelet``.
+    boundary : str, optional
+        The boundary condition to be imposed at the edges of the input signal ``x``.
+        Allowed values are ``"mirror"``, ``"zeros"``, and ``"periodic"``. Default is ``"mirror"``.
+    time_axis : int, optional
+        Axis on which the time is defined for input ``x`` (default is last, or -1). Note that the time axis of the
+        wavelets must be last.
+    freq_axis : int, optional
+        Axis of ``wavelet`` for the frequencies (default is second or 1)
+    order_axis : int, optional
+        Axis of ``wavelet`` for the orders (default is first or 0)
+
+    Returns
+    -------
+    wtx : np.ndarray
+        Time-domain wavelet transform of input ``x``. The axes of ``wtx`` will be organized as (x axes), orders, frequencies, time
+        unless ``time_axis`` is different from last (-1) in which case it will be moved back to its original position within the axes of ``x``.
+
+    Examples
+    --------
+
+    Apply a wavelet transform with a Morse wavelet with gamma parameter 3, beta parameter 4, at radian frequency 0.2 cycles per unit time:
+
+    >>> x = np.random.random(1024)
+    >>> wavelet, _ = morse_wavelet(1024, 3, 4, np.array([2*np.pi*0.2]))
+    >>> wtx = wavelet_transform(x, wavelet)
+
+    The input signal can have an arbitrary number of dimensions but its ``time_axis`` must be specified if it is not the last:
+
+    >>> x = np.random.random((1024,10,15))
+    >>> wavelet, _ = morse_wavelet(1024, 3, 4, np.array([2*np.pi*0.2]))
+    >>> wtx = wavelet_transform(x, wavelet,time_axis=0)
+
+    Raises
+    ------
+    ValueError
+        If the time axis is outside of the valid range ([-1, N-1]).
+        If the shape of time axis is different for input signal and wavelet.
+        If boundary optional argument is not in ["mirror", "zeros", "periodic"]``.
+
+    See Also
+    --------
+    :func:`morse_wavelet`, `morse_wavelet_transform`, `morse_freq`
+    """
+    # time_axis must be in valid range
+    if time_axis < -1 or time_axis > len(x.shape) - 1:
+        raise ValueError(
+            f"time_axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(x.shape) - 1}])."
+        )
```

Comment:
I think it needs a `raise ValueError` here.

Reply:
Don't get it?